### ISSUE TITLE
Numeric Instability in cuda for WGAN Gradient Penalty

### ISSUE THREAD

**Issue by AliengirlLiv** [pytorch/pytorch#16433]
## 🐛 Bug

I'm implementing a Wasserstein GAN model in Pytorch. The code is a re-implementation of [this tensorflow code](https://github.com/adler-j/bwgan). That code is a bit of a generalized wasserstein gan, but if you set the hyperparameters appropriately it reduces to a normal wasserstein gan. The re-implementation was done closely in that if we compute the forward output of the generator/discriminator for fixed images/noise while settings all weights the same, we get the same output in both tensorflow and pytorch.

All of my pytorch code can be found in [this jupyter notebook](https://github.com/AliengirlLiv/gan-image-metrics/blob/eval/BWGAN-PyTorch.ipynb), . The only dependencies for the notebook are pytorch (nightly), matplotlib, and numpy. There is a variable in the last cell called use_cuda. If that variable is set to true, then the discriminator's gradient loss penalty term blows up to infinity.  If it is set to false the loss stays reasonable. So it looks like one of the operations is numerically unstable on cuda but not on pytorch. Also, the model was unstable on cpu on pytorch 1.0. So whatever operation is numerically unstable looks like it got improved on cpu between 1.0 and the current nightly.

 - PyTorch Version (e.g., 1.0.0.dev20190125):
 - OS (e.g., Linux): Linux
 - How you installed PyTorch (`conda`, `pip`, source): pip
 - Python version: 3.6
 - CUDA/cuDNN version:  10.0

**Comment by yf225** [comment]
cc @soumith

**Comment by AliengirlLiv** [comment]
It seems probable that the issue is related to issue #16532  with calling `torch.autograd.grad` within a function while using multiple GPUs.  We tried running the code on 1 GPU and it worked.  The instability only appears with multiple GPUs.

**Comment by aluo-x** [comment]
Also suffering with this issue.

Pytorch 1.0.1
OS: Ubuntu 16.04.6 LTS
Pytorch built against CUDA 10.1
Python version: 3.6.7

**Comment by pietern** [comment]
I think this is the same as #19024.

**Comment by pietern** [comment]
This is also the same as #16532 it looks like. Closing this issue to consolidate. Even though this issue is older, the other one has more discussion about the underlying cause and possible solutions.
